src/algebra/star/basic.lean

```diff
@@ -0,0 +1,152 @@
+Authors: Scott Morrison.
+-/
+import tactic.apply_fun
+import algebra.ordered_ring
+import algebra.big_operators.basic
+import data.equiv.ring
+
+/-!
+# star_monoid, star_ring
+
+We introduce the basic algebraic notions of star monoids, and star rings.
+Star algebras are introduced in `algebra.algebra.star`.
+
+These are implemented as "mixin" typeclasses, so to summon a star ring (for example)
+one needs to write `(R : Type) [ring R] [star_ring R]`.
+This avoids difficulties with diamond inheritance.
+
+For now we simply do not introduce notations,
+as different users are expected to feel strongly about the relative merits of
+`r^*`, `r†`, `rᘁ`, and so on.
+
+Our star rings are actually star semirings, but of course we can prove
+`star_neg : star (-r) = - star r` when the underlying semiring is a ring.
+-/
+
+
+universes u v
+
+/--
+Notation typeclass (with no default notation!) for an algebraic structure with a star operation.
+-/
+class has_star (R : Type u) :=
+(star : R → R)
+
+variables {R : Type u}
+
+/--
+A star operation (e.g. complex conjugate).
+-/
+def star [has_star R] (r : R) : R := has_star.star r
+
+/--
+Typeclass for a star operation with is involutive.
+-/
+class has_involutive_star (R : Type u) extends has_star R :=
+(star_involutive : function.involutive star)
+
+@[simp] lemma star_star [has_involutive_star R] (r : R) : star (star r) = r :=
+has_involutive_star.star_involutive _
+
+lemma star_injective [has_involutive_star R] : function.injective (star : R → R) :=
+has_involutive_star.star_involutive.injective
+
+/--
+A *-monoid is a monoid `R` with an involutive operations `star`
+so `star (r * s) = star s * star r`.
+-/
+class star_monoid (R : Type u) [monoid R] extends has_involutive_star R :=
+(star_mul : ∀ r s : R, star (r * s) = star s * star r)
+
+@[simp] lemma star_mul [monoid R] [star_monoid R] (r s : R) : star (r * s) = star s * star r :=
+star_monoid.star_mul r s
+
+variables (R)
+
+@[simp] lemma star_one [monoid R] [star_monoid R] : star (1 : R) = 1 :=
+begin
+  have w : ∀ r, (star 1 : R) * r = r,
+  { intro r, apply star_injective, simp, },
+  simpa using w 1,
+end
+
+variables {R}
+
+/-- `star` as an `mul_equiv` from `R` to `Rᵒᵖ` -/
+def star_mul_equiv [semiring R] [star_monoid R] : R ≃* Rᵒᵖ :=
+{ to_fun := λ x, opposite.op (star x),
+  map_mul' := λ x y, (star_mul x y).symm ▸ (opposite.op_mul _ _),
+  ..(has_involutive_star.star_involutive.to_equiv star).trans opposite.equiv_to_opposite}
+
+/--
+A *-ring `R` is a (semi)ring with an involutive `star` operation which is additive
+which makes `R` with its multiplicative structure into a *-monoid
+(i.e. `star (r * s) = star s * star r`).
+-/
+class star_ring (R : Type u) [semiring R] extends star_monoid R :=
+(star_zero : star 0 = 0)
+(star_add : ∀ r s : R, star (r + s) = star r + star s)
+
+variables (R)
+
+@[simp] lemma star_zero [semiring R] [star_ring R] : star (0 : R) = 0 :=
+star_ring.star_zero
+
+variables {R}
+
```

Comment:
Strictly `star_zero` is no longer needed - it can be obtained as `star_add_equiv.map_zero`. Should I remove it?